tracr/craft/bases.py

```diff
@@ -135,48 +163,96 @@ def project(
     if isinstance(basis, VectorSpaceWithBasis):
       basis = basis.basis
     components = []
+    # use a set for lookups, greatly improves performance
+    basis_set = self.get_basis_set()
+    direction_to_index = self.get_direction_to_index()
     for direction in basis:
-      if direction in self.basis_directions:
+      if direction in basis_set:
         components.append(
-            self.magnitudes[..., self.basis_directions.index(direction)])
+            self.magnitudes[..., direction_to_index[direction]])
       else:
         components.append(np.zeros_like(self.magnitudes[..., 0]))
-    return VectorInBasis(list(basis), np.stack(components, axis=-1))
-
-
+    
+    if isinstance(basis, VectorSpaceWithBasis): # we can save compute and construct the vector in basis from the vector space
```

Comment:
comment should clarify what's the condition that we can do this.  Specifically basis is VectorSpaceWithBasis as opposed to what else?

Also I find it confusing the basis is a VectorSpaceWithBasis object sometimes, shouldn't this be called differently?

Reply:
my bad I had changed the name of basis in an older versions and distinguished between the parameter and just a list of basis directions so the source VectorSpaceWithBasis object can be maintained.
I've now named the parameter `basis_container` and if it is a VectorSpaceWithBasis we extract the `basis_directions` otherwise we cast a sequence of basis directions to `basis_directions`. Now we can guarantee `basis_directions` is a sequence of basis directions while leaving a VectorSpaceWithBasis (`basis_container`) alone so we can cast straight to a VectorInBasis